### ISSUE TITLE
Use a query as table (subquery)

### ISSUE THREAD

**Issue by that0n3guy** [metabase/metabase#1822]
Chart.io has a very powerful feature that allows you to create "custom tables".  Basically, you enter in any sql you want and the result is your new table.  You can then set the metadata for the columns that the query returns.

So if your query (lets call it "Custom1" table) returns 3 columns, you would now see a new table, called "Custom1" in your metabase admin.  You can then set the metadata for that table (visibility, type, and details) for those 3 columns.  Then you can use your new "Custom1" table when asking questions in metabase.

Mysql (I think general SQL) allows you to use a subquery as your `FROM` so this is easy form a sql standpoint.

This would be very handy for:
- Creating complex queries but still letting the lay-business person slice and dice it.
- Easily to joins multiple depths
- limit the columns shown an a normal select query
- join tables that have funky polymorphic relationships (I have lots of those).

Currently, something as simple as this:

```
select
    FROM_UNIXTIME(A.createDate) 'createDate',
    count(C.id)
from x2_contacts as C
    inner join x2_actions A ON C.id = A.associationId
where A.associationType = 'contacts'
    and A.type = 'workflow'
group by MONTH(FROM_UNIXTIME(A.createDate))
```

... is has to be done by custom sql because my x2_actions table is polymorphic (see the `where A.associationType = 'contacts'`).   Having "virtual" or "custom" tables is a decent way to fix this and allow for a LOT of flexibility.

**Comment by camsaul** [comment]
This would probably be pretty feasible on the backend for SQL DBs. Mongo + Druid would be trickier. The biggest challenge here I think would be the UI

**Comment by tlrobinson** [comment]
+1 I agree this would be super cool.

As a workaround for now you can create SQL views which will show up as tables in Metabase (I've confirmed it works in MySQL and Postgres, haven't checked the others).

e.x. `CREATE VIEW view_name AS SELECT ...`

Make sure to sync the database ("Sync" button in `/admin/databases/:id`) after creating the view.

**Comment by salsakran** [comment]
@camsaul think this + a way to mark recurringly useful queries could be a pretty good alternative to virtual tables.

**Comment by tlrobinson** [comment]
👍  This basically *is* virtual tables, with the benefit that if we add all the features we planned for virtual tables to the query builder instead then you can use those features in normal questions as well as queries marked as a "virtual table" (or whatever we call it)

**Comment by LukeAbell** [comment]
+1 💯

**Comment by marcoruggine** [comment]
+1
I really need this feature! Using the query builder on views would be a game changer for us

**Comment by j005u** [comment]
+1
@camsaul Please don't skip out on this in terms of mongo. It's literally just appending stages to the aggregation pipeline. This would make everything so much simpler for us (building a multi tenant self service portal for micropayments ontop of metabase) even without mongo views, which haven't been deployed to most DB hosting solutions yet. Most other DBs can also solve part of this issue with regular views, mongo < 3.4 can't.

In terms of UI, what's the complication? I'd just like to select any other query I have saved as the input (Slamdata does something like this too) and then build on top of that either with the UI or manually. The rest should pretty much be concating the two arrays together?

One more note: Slamdata has these virtual views and also enables you to cache the output of any view like this in order to run subsequent queries faster. Just a thought.

**Comment by timabe** [comment]
+1 

I find myself often wanting to create a "subtable" that other users of metabase can play around with using the query builder. 

One idea for the UI: You could add it to the ellipsis on the right side of the query builder. When you click that, you'd see an additional option
![screen shot 2017-02-03 at 4 09 57 pm](https://cloud.githubusercontent.com/assets/1427277/22613348/70056580-ea2c-11e6-84e4-0001756d3346.png)
If you click it, you get a new query builder on top of the one you had before. I don't love the second part of the design, but I think the first part is okay.
![screen shot 2017-02-03 at 4 15 51 pm](https://cloud.githubusercontent.com/assets/1427277/22613367/938ab74e-ea2c-11e6-99f5-145b74efaa5a.png)

**Comment by marszall87** [comment]
Great idea, I'm really looking forward to use this. How about just allowing queries to be run on the results of another query? You can throw a result set from any query to a temporary table in SQLite in memory or smth and then query that table. Changes in UI would be rather simple, not sure what about the metadata though.

Oh, and it would allow subqueries on non-SQL data sources, like Google Analytics - I could really use this, as GA API is very limited in terms of grouping data and calculating more advanced metrics.

**Comment by mazameli** [comment]
In order to include a new option in the query builder's data selector to allow choosing saved questions, I'm thinking we can use the pattern we're using for DBs with schemas:

![saved questions](https://cloud.githubusercontent.com/assets/2223916/23929291/41de8eaa-08e2-11e7-9cc2-bfaa21560811.png)

This would open up the collections list, in a similar way that we're doing within the Pulse creation flow:

![list of collections](https://cloud.githubusercontent.com/assets/2223916/23929306/5e2611fa-08e2-11e7-900e-08e249da5144.png)

Selecting a collection would drill into it, displaying the list of questions:
![list of questions](https://cloud.githubusercontent.com/assets/2223916/23929311/662ccf4c-08e2-11e7-988a-3002fe92b34d.png)

If no collections exist, we can get rid of the down arrow/chevron on the `Saved Questions` menu item, and skip straight to the list of questions upon clicking `Saved Questions`.

**Comment by salsakran** [comment]
would it make sense to have collections be equivalent to schemas? i.e. "under" the `Saved Questions` faux-db.

**Comment by luissancho** [comment]
That's a really great solution @mazameli, I believe it will be a huge step forward for Metabase as a product.

**Comment by si14** [comment]
Do you think it's possible to make the feature avaiable in SQL mode, too?

**Comment by goodfella1408** [comment]
If this feature is extensible to NOSQL DB's it would really make metabase stand apart from other BI tools (including commercial ones like Looker, Sisense,BIME) .But for that a temp storage (ideally schemaless like mongo or elasticsearch and with efficient deletes and aggregation) must be used .One of the challenges with such a approach is to ensure multitenancy in the temp store wrt different filters that can be applied on the original question

**Comment by camsaul** [comment]
Implemented by #4613

**Comment by amywtlin** [comment]
+1 for also in SQL mode. 

I tried using one saved question as raw data but found out I can't aggregate the columns handled with `case when...` statement. It's really a pity and I was very close to make full use of it!

**Comment by popwebz** [comment]
This does not seem to be implemented for native SQL queries that are saved questions
This would be awesome.

**Comment by mazameli** [comment]
@amywtlin @popwebz Thanks for the feedback. I think what you're proposing is basically what's outlined in this issue: https://github.com/metabase/metabase/issues/1825

**Comment by GrossBetruger** [comment]
+1

**Comment by camsaul** [comment]
@GrossBetruger this was implemented by #4613.
